open-telemetry/opentelemetry-python
# Make event attributes immutable

# Description

This PR makes it more difficult for someone to edit the attributes of an `Event` once it has been added to a `Span`, as well as adding some regression tests around the other properties of an `Event` to ensure they have similar protections.

It isn't particularly straightforward or idiomatic to make objects truly immutable in Python, but if these steps don't go far enough let me know!

I don't believe a documentation change is necessary because events are meant to be immutable[ according to the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#add-events), but this should probably be considered a breaking change in case people were relying on the event attributes being mutable up until now.

Fixes #1038

## Type of change

Please delete options that are not relevant.


# How Has This Been Tested?

I've added regression tests to:
- make sure that the properties of an `Event` raise an error when someone directly tries to change them; and
- make sure that an error is raised if someone tries to directly edit any of the `attributes` dictionary tuples.

I've also added a new test for my change to make sure that an error is raised if someone tryies to directly edit the `attributes` dictionary itself.

# Checklist:
